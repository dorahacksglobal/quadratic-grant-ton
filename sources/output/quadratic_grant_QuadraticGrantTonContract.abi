{"name":"QuadraticGrantTonContract","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"ChangeOwner","header":256331011,"fields":[{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"Project","header":null,"fields":[{"name":"id","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"pid","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"area","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"votes","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"contribution","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"voter_number","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"voters","type":{"kind":"dict","key":"address","value":"int"}}]},{"name":"Votes","header":null,"fields":[{"name":"project_id","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"votes","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"Round","header":null,"fields":[{"name":"id","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"tax_adjustment_multiplier","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"voting_unit","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"status","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"fund","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"project_number","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"total_area","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"contribution","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"SetFund","header":2162894902,"fields":[{"name":"fund","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"BatchUploadProjects","header":2067884959,"fields":[{"name":"count","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"projects","type":{"kind":"dict","key":"int","value":"int"}}]},{"name":"Vote","header":3951512617,"fields":[{"name":"project_id","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"votes","type":{"kind":"simple","type":"int","optional":false,"format":257}}]}],"receivers":[{"receiver":"internal","message":{"kind":"typed","type":"SetFund"}},{"receiver":"internal","message":{"kind":"typed","type":"BatchUploadProjects"}},{"receiver":"internal","message":{"kind":"typed","type":"Vote"}},{"receiver":"internal","message":{"kind":"text","text":"EndRound"}},{"receiver":"internal","message":{"kind":"typed","type":"Deploy"}}],"getters":[{"name":"roundInfo","arguments":[],"returnType":{"kind":"simple","type":"Round","optional":false}},{"name":"project","arguments":[{"name":"project_id","type":{"kind":"simple","type":"int","optional":false,"format":257}}],"returnType":{"kind":"simple","type":"Project","optional":true}},{"name":"owner","arguments":[],"returnType":{"kind":"simple","type":"address","optional":false}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"137":{"message":"Masterchain support is not enabled for this contract"},"14329":{"message":"Voting amount not integer"},"15518":{"message":"Amount is too small"},"26909":{"message":"Round status is invalid"},"51754":{"message":"Insufficient funds"}},"interfaces":["org.ton.introspection.v0","org.ton.abi.ipfs.v0","org.ton.deploy.lazy.v0","org.ton.debug.v0","org.ton.chain.workchain.v0","org.ton.ownable"]}