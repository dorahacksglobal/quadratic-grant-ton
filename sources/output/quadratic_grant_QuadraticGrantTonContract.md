# TACT Compilation Report
Contract: QuadraticGrantTonContract
BOC Size: 1948 bytes

# Types
Total Types: 12

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## ChangeOwner
TLB: `change_owner#0f474d03 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{newOwner:address}`

## Project
TLB: `_ id:int257 pid:int257 area:int257 votes:int257 contribution:int257 voter_number:int257 voters:dict<address, int> = Project`
Signature: `Project{id:int257,pid:int257,area:int257,votes:int257,contribution:int257,voter_number:int257,voters:dict<address, int>}`

## Votes
TLB: `_ project_id:int257 votes:int257 = Votes`
Signature: `Votes{project_id:int257,votes:int257}`

## Round
TLB: `_ id:int257 tax_adjustment_multiplier:int257 voting_unit:int257 status:int257 fund:int257 project_number:int257 total_area:int257 contribution:int257 = Round`
Signature: `Round{id:int257,tax_adjustment_multiplier:int257,voting_unit:int257,status:int257,fund:int257,project_number:int257,total_area:int257,contribution:int257}`

## SetFund
TLB: `set_fund#80eb2836 fund:int257 = SetFund`
Signature: `SetFund{fund:int257}`

## BatchUploadProjects
TLB: `batch_upload_projects#7b416b9f count:int257 projects:dict<int, int> = BatchUploadProjects`
Signature: `BatchUploadProjects{count:int257,projects:dict<int, int>}`

## Vote
TLB: `vote#eb874c29 project_id:int257 votes:int257 = Vote`
Signature: `Vote{project_id:int257,votes:int257}`

# Get Methods
Total Get Methods: 3

## roundInfo

## project
Argument: project_id

## owner

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
137: Masterchain support is not enabled for this contract
14329: Voting amount not integer
15518: Amount is too small
26909: Round status is invalid
51754: Insufficient funds