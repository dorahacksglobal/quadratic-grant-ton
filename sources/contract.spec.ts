import { Dictionary, toNano, SendMode } from "ton";
import { ContractSystem } from "@tact-lang/emulator";
import { QuadraticGrantTonContract, Votes } from "./output/quadratic_grant_QuadraticGrantTonContract";

describe("contract", () => {
    it("should process correctly", async () => {
        // Create ContractSystem and deploy contract
        let system = await ContractSystem.create();

        let owner = system.treasure("owner");
        let nonOwner = system.treasure("non-owner");
        let contract = system.open(await QuadraticGrantTonContract.fromInit(owner.address, 1n, 10n, 10n, 4000n));
        system.name(contract.address, "contract");

        // This object would track all transactions in this contract
        let track = system.track(contract);
        // This object would track all logs
        let logger = system.log(contract.address);

        // ============ STEP 1: Deploy contract ============
        await contract.send(owner, { value: toNano(1) }, { $$type: "Deploy", queryId: 0n });
        await system.run();
        expect(track.collect()).toMatchSnapshot(); // 1
        expect(await contract.getBalance()).toMatchInlineSnapshot(`0n`); // 2

        // ================== STEP 3: Upload projects ==================
        await contract.send(
            owner,
            {
                value: toNano("0.1"),
            },
            {
                $$type: "BatchUploadProjects",
                count: 2n,
                projects: Dictionary.empty<bigint, bigint>().set(1n, 1n).set(2n, 2n),
            }
        );
        await contract.send(
            owner,
            {
                value: toNano("0.1"),
            },
            {
                $$type: "BatchUploadProjects",
                count: 1n,
                projects: Dictionary.empty<bigint, bigint>().set(1n, 3n),
            }
        );
        await system.run();
        expect(track.collect()).toMatchSnapshot(); // 3
        expect(await contract.getRoundInfo()).toMatchSnapshot(); // 4
        expect(await contract.getBalance()).toMatchInlineSnapshot(`0n`); // 5

        // ================== STEP 4: Vote ==================
        await contract.send(
            nonOwner,
            { value: toNano("1.6") }, // 1.600000000
            {
                $$type: "Vote",
                project_id: 1n,
                votes: toNano("1.6"),
            }
        );
        await contract.send(
            nonOwner,
            { value: toNano("0.9") },
            {
                $$type: "Vote",
                project_id: 1n,
                votes: toNano("0.9"),
            }
        );
        await contract.send(
            nonOwner,
            { value: toNano("1.6") },
            {
                $$type: "Vote",
                project_id: 2n,
                votes: toNano("1.6"),
            }
        );
        await contract.send(
            owner,
            { value: toNano("0.9") },
            {
                $$type: "Vote",
                project_id: 1n,
                votes: toNano("0.9"),
            }
        );
        // logger.reset();
        await system.run();
        // console.log(logger.collect());
        expect(track.collect()).toMatchSnapshot(); // 6
        expect(await contract.getRoundInfo()).toMatchSnapshot(); // 7
        expect(await contract.getBalance()).toMatchInlineSnapshot(`4925526000n`); // 8

        // ================== STEP 5: Upload More Projects==================
        await contract.send(
            owner,
            {
                value: toNano("0.1"),
            },
            {
                $$type: "BatchUploadProjects",
                count: 1n,
                projects: Dictionary.empty<bigint, bigint>().set(1n, 3n),
            }
        );

        await system.run();
        expect(track.collect()).toMatchSnapshot(); // 9
        expect(await contract.getBalance()).toMatchInlineSnapshot(`4925526000n`); // 10
        expect(await contract.getProject(1n)).toMatchSnapshot(); // 11
        expect(await contract.getProject(2n)).toMatchSnapshot(); // 12
        expect(await contract.getProject(3n)).toMatchSnapshot(); // 13

        // ================== STEP 5: End Round ==================
        await contract.send(owner, { value: toNano("1") }, "EndRound");
        await system.run();
        expect(track.collect()).toMatchSnapshot(); // 14
        expect(await contract.getBalance()).toEqual(0n);
    });
});
