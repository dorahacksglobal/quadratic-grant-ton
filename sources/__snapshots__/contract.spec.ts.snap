// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract should process correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{946A98B60000000000000000}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@contract",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11111n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{AFF90F570000000000000000}",
              "type": "cell",
            },
            "bounce": true,
            "from": "@contract",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 987693000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should process correctly 3`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{7B416B9F00000000000000000000000000000000000000000000000000000000000000016_}
 x{CFF}
  x{50000000000000000000000000000000000000000000000000000000000000000C_}
  x{400000000000000000000000000000000000000000000000000000000000000014_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@contract",
          "type": "internal",
          "value": 100000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 18369n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": true,
            "from": "@contract",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 80631000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{7B416B9F0000000000000000000000000000000000000000000000000000000000000000E_}
 x{A0200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001C_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@contract",
          "type": "internal",
          "value": 100000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 15649n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": true,
            "from": "@contract",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 83351000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should process correctly 4`] = `
{
  "$$type": "Round",
  "contribution": 0n,
  "fund": 4000n,
  "id": 1n,
  "project_number": 3n,
  "status": 0n,
  "tax_adjustment_multiplier": 10n,
  "total_area": 0n,
  "voting_unit": 10n,
}
`;

exports[`contract should process correctly 6`] = `
[
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{EB874C2900000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000017D784002_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(non-owner)",
          "to": "@contract",
          "type": "internal",
          "value": 1600000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 17785n,
      },
    ],
  },
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{EB874C290000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000D693A402_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(non-owner)",
          "to": "@contract",
          "type": "internal",
          "value": 900000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 19563n,
      },
    ],
  },
  {
    "$seq": 5,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{EB874C2900000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000017D784002_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(non-owner)",
          "to": "@contract",
          "type": "internal",
          "value": 1600000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 18410n,
      },
    ],
  },
  {
    "$seq": 6,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{EB874C290000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000D693A402_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@contract",
          "type": "internal",
          "value": 900000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 18716n,
      },
    ],
  },
]
`;

exports[`contract should process correctly 7`] = `
{
  "$$type": "Round",
  "contribution": 5000000000n,
  "fund": 4000n,
  "id": 1n,
  "project_number": 3n,
  "status": 0n,
  "tax_adjustment_multiplier": 10n,
  "total_area": 120n,
  "voting_unit": 10n,
}
`;

exports[`contract should process correctly 9`] = `
[
  {
    "$seq": 7,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{7B416B9F0000000000000000000000000000000000000000000000000000000000000000E_}
 x{A0200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001C_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@contract",
          "type": "internal",
          "value": 100000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 15049n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": true,
            "from": "@contract",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 83951000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should process correctly 11`] = `
{
  "$$type": "Project",
  "area": 80n,
  "contribution": 3400000000n,
  "id": 1n,
  "pid": 1n,
  "voter_number": 2n,
  "voters": Dictionary {
    "_key": {
      "bits": 267,
      "parse": [Function],
      "serialize": [Function],
    },
    "_map": Map {
      "a:EQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3LUF" => 9n,
      "a:EQCVnZ1On-Ja4xfAfMbsq--jatb5sNnOUN421AHaXbebcJ4N" => 25n,
    },
    "_value": {
      "parse": [Function],
      "serialize": [Function],
    },
  },
  "votes": 34n,
}
`;

exports[`contract should process correctly 12`] = `
{
  "$$type": "Project",
  "area": 40n,
  "contribution": 1600000000n,
  "id": 2n,
  "pid": 2n,
  "voter_number": 1n,
  "voters": Dictionary {
    "_key": {
      "bits": 267,
      "parse": [Function],
      "serialize": [Function],
    },
    "_map": Map {
      "a:EQCVnZ1On-Ja4xfAfMbsq--jatb5sNnOUN421AHaXbebcJ4N" => 16n,
    },
    "_value": {
      "parse": [Function],
      "serialize": [Function],
    },
  },
  "votes": 16n,
}
`;

exports[`contract should process correctly 13`] = `
{
  "$$type": "Project",
  "area": 0n,
  "contribution": 0n,
  "id": 3n,
  "pid": 3n,
  "voter_number": 0n,
  "voters": Dictionary {
    "_key": {
      "bits": 267,
      "parse": [Function],
      "serialize": [Function],
    },
    "_map": Map {},
    "_value": {
      "parse": [Function],
      "serialize": [Function],
    },
  },
  "votes": 0n,
}
`;

exports[`contract should process correctly 14`] = `
[
  {
    "$seq": 8,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "EndRound",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(owner)",
          "to": "@contract",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11182n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": true,
            "from": "@contract",
            "to": "@treasure(owner)",
            "type": "internal",
            "value": 5913344000n,
          },
        ],
      },
    ],
  },
]
`;
